File: prototype/rails/src/app/domain/listings/listing_stock.rb
require_relative "listing_status"
require_relative "stock_change"

module Domain
  module Listings
    # What a listing has left once an order takes stock or hands it back.
    ListingStock = Data.define(:quantity, :status) do
      def self.after(change, quantity:, status:, items:)
        case change
        when StockChange::TAKE then after_sale(quantity: quantity, status: status, sold: items)
        when StockChange::RESTORE then after_restock(quantity: quantity, status: status, restored: items)
        when StockChange::KEEP then new(quantity: quantity, status: status)
        else raise ArgumentError, "unknown stock change: #{change.inspect}"
        end
      end

      def self.after_sale(quantity:, status:, sold:)
        reject_an_empty_change(sold)
        raise ArgumentError, "a listing that is #{status} cannot be sold" unless status == ListingStatus::FOR_SALE
        raise ArgumentError, "a listing with #{quantity} left cannot sell #{sold}" if sold > quantity

        remaining = quantity - sold
        new(quantity: remaining, status: remaining.zero? ? ListingStatus.transition(status, ListingStatus::SOLD) : status)
      end

      def self.after_restock(quantity:, status:, restored:)
        reject_an_empty_change(restored)

        new(
          quantity: quantity + restored,
          status: status == ListingStatus::SOLD ? ListingStatus.transition(status, ListingStatus::FOR_SALE) : status
        )
      end

      def self.reject_an_empty_change(items)
        raise ArgumentError, "a stock change covers at least one item, got #{items}" if items < 1
      end
      private_class_method :reject_an_empty_change
    end
  end
end

File: prototype/node/src/app/core/listings/listing-stock.ts
import { transitionListing, type ListingStatus } from './listing-status.ts'
import type { StockChange } from './stock-change.ts'

export type ListingStock = { quantity: number; status: ListingStatus }

function rejectAnEmptyChange(items: number): void {
  if (items < 1) {
    throw new RangeError(`a stock change covers at least one item, got ${items}`)
  }
}

export function stockAfterSale(input: { quantity: number; status: ListingStatus; sold: number }): ListingStock {
  const { quantity, status, sold } = input
  rejectAnEmptyChange(sold)
  if (status !== 'for_sale') {
    throw new RangeError(`a listing that is ${status} cannot be sold`)
  }
  if (sold > quantity) {
    throw new RangeError(`a listing with ${quantity} left cannot sell ${sold}`)
  }

  const remaining = quantity - sold
  return { quantity: remaining, status: remaining === 0 ? transitionListing(status, 'sold') : status }
}

export function stockAfterRestock(input: { quantity: number; status: ListingStatus; restored: number }): ListingStock {
  const { quantity, status, restored } = input
  rejectAnEmptyChange(restored)

  return {
    quantity: quantity + restored,
    status: status === 'sold' ? transitionListing(status, 'for_sale') : status,
  }
}

export function stockAfter(
  change: StockChange,
  input: { quantity: number; status: ListingStatus; items: number },
): ListingStock {
  const { quantity, status, items } = input
  switch (change) {
    case 'take':
      return stockAfterSale({ quantity, status, sold: items })
    case 'restore':
      return stockAfterRestock({ quantity, status, restored: items })
    case 'keep':
      return { quantity, status }
    default:
      throw new TypeError(`unknown stock change: ${JSON.stringify(change)}`)
  }
}
